### STRINGS: Cadenas de caracteres
- Cadenas de caracteres.
- Operaciones sobre strings (``str``, ``count``, ``isalpha``/``isdigit``, ``upper``/``lower``, ``strip``, ``split``, ``join``)
- Operaciones entre strings (concatenar ``+``, repeticion ``*`` y lógicas ``<``, ``>``)
- Funciones a nivel de carácter (``ord`` y ``char``)
- Slices (``<secuencia>[:]``)

#### CADENAS DE CARACTERES
- A efectos prácticos son iguales que las tuplas
- Representadas internamente como una secuencia de bytes unicode, por lo que **aceptan cualquier carácter unicode**.

In [83]:
print("Cadena de caracteres con comillas dobles")
print('Cadena de caracteres con comillas simples')
print('Cadena con comillas "anidadas" en su contenido')
print(""" Cadena
multilínea""")
print('Bonus:' ',separar cadenas con espacio,' 'las concatena')

Cadena de caracteres con comillas dobles
Cadena de caracteres con comillas simples
Cadena con comillas "anidadas" en su contenido
 Cadena
multilínea
Bonus:,separar cadenas con espacio,las concatena


#### OPERACIONES BÁSICAS SOBRE STRINGS
``str``: crea una cadena del objeto indicado

In [84]:
str((1,2,3,4,5))

'(1, 2, 3, 4, 5)'

``<cadena>.count``: devuelve el numero de apariciones **no solapadas** de una subcadena

In [85]:
print('baba yagá'.count('ba'))
print('baba yagá'.count('a'))

2
3


``<cadena>.isalpha()`` / ``<cadena>.isdigit()``: booleano que indica si **todos** los caracteres de la cadena son alfabéticos o numéricos

In [86]:
cadena_alpha = 'alfabetico'
cadena_numer = '41596132'
print(cadena_alpha.isalpha(), cadena_alpha.isdigit())
print(cadena_numer.isalpha(), cadena_numer.isdigit())

True False
False True


``<cadena>.upper()`` / ``<cadena>.lower()``: crea una copia de la cadena en mayúscula o minúscula

In [87]:
cadena = 'Tutti frutti'
print(cadena.upper())
print(cadena.lower())

TUTTI FRUTTI
tutti frutti


``<cadena>.strip()``: devuelve una copia de la cadena **eliminando los espacios (chars) del principio/final**

In [88]:
'      espaciado  quitado '.strip()

'espaciado  quitado'

``<cadena>.split()``: crea una lista con los caraceteres de la cadena separados por un separador. Si no se especifica, por defecto el separador es un char

In [89]:
cadena_char = '1 2 3 4'
secuencia = cadena_char.split()
print(secuencia, type(secuencia)) 
cadena_barr = '1|2|3|4'
secuencia = cadena_barr.split('|')  # hay que indicar cuál es el separador
print(secuencia, type(secuencia))

['1', '2', '3', '4'] <class 'list'>
['1', '2', '3', '4'] <class 'list'>


``<cadena>.join``: hace el paso inverso a split. Parte de una secuencia y convierte sus elementos en una cadena de caracteres

In [90]:
secuencia = ('1','char_2','3','char_4')
cadena = ', '.join(secuencia)    # hay que indicar qué separador los une, normalmente comas y espacios 
cadena, type(cadena)

('1, char_2, 3, char_4', str)

#### OPERACIONES ENTRE STRINGS
Estas operaciones devuelven una nueva cadena.
- Concatenación (``+``)
- Repetición (``*``)
- Lógicas (``<``, ``>``): Comparan los valores numéricos de los caracteres, que están ordenados lexicográficamente.

In [91]:
saludo = 'hola'
despedida = " y adios"
total = 3*saludo + despedida
print(total)
print(saludo > total, saludo > despedida)

holaholahola y adios
False True


#### FUNCIONES A NIVEL DE CARÁCTER
- ``ord(c)``: devuelve la representación numérica del carácter *c*
- ``char(n)``: devuelve el caracter de la representación numérica *n*

In [92]:
ord('h'), ord(' '), chr(128013)

(104, 32, '🐍')

Por eso, antes daba que ``saludo`` es mayor que ``despedida``, porque ``'h'`` es mayor que ``' '`` (104 > 32)

##### SLICE
Operación mediante la cual se extraen elementos de una secuencia.
- Notación: ``secuencia[<start>:<stop>:<step>]``
- Estos tres valores son opcionales y pueden ser negativos
- Se generan copias, por lo que los cambios sobre el *slice* **no alteran al objeto original**

In [93]:
l = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
print('a', l[1:7:2])
print('b', l[-1:-7:-2])  # -1 (empieza desde el 9) : -7 (termina en el 3, sin incluir) : -2 (recorrido inverso de 2 en 2)
print('c', l[1:7])
print('d', l[:])
print('e', l[:-2])
print('f', l[::3])   # todos los elementos desde el principio hasta el final con un step = 3

a [1, 3, 5]
b [9, 7, 5]
c [1, 2, 3, 4, 5, 6]
d [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
e [0, 1, 2, 3, 4, 5, 6, 7]
f [0, 3, 6, 9]


##### USOS TÍPICOS
- Copiar una cadena o secuencia
- Invertir una cadena o secuencia
- Modificar una lista

In [94]:
secuencia = 'War does not determine who is right; only who is left'
copia = secuencia[:]
inversion = secuencia[::-1]
print(copia)
print(inversion)
print(l[::-3])


War does not determine who is right; only who is left
tfel si ohw ylno ;thgir si ohw enimreted ton seod raW
[9, 6, 3, 0]


In [95]:
l[3:7] = [99, 99, 99, 99]
print(l)
l[5:] = [42]
print(l)
l[:5] = [24]
print(l)

[0, 1, 2, 99, 99, 99, 99, 7, 8, 9]
[0, 1, 2, 99, 99, 42]
[24, 42]
